agent/consul/rpc_client.go

```diff
@@ -0,0 +1,101 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/metadata"
+	"github.com/hashicorp/consul/agent/pool"
+	"github.com/hashicorp/consul/tlsutil"
+	"google.golang.org/grpc"
+)
+
+const (
+	grpcBasePath = "/consul"
+)
+
+func dialGRPC(addr string, _ time.Duration) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = conn.Write([]byte{pool.RPCGRPC})
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+type RPCClient struct {
+	rpcPool   *pool.ConnPool
+	grpcConns sync.Map
+	logger    *log.Logger
+}
+
+func NewRPCClient(logger *log.Logger, config *Config, tlsConfigurator *tlsutil.Configurator, maxConns int, maxIdleTime time.Duration) *RPCClient {
+	return &RPCClient{
+		rpcPool: &pool.ConnPool{
+			SrcAddr:    config.RPCSrcAddr,
+			LogOutput:  config.LogOutput,
+			MaxTime:    maxIdleTime,
+			MaxStreams: maxConns,
+			TLSWrapper: tlsConfigurator.OutgoingRPCWrapper(),
+			ForceTLS:   config.VerifyOutgoing,
+		},
+		logger: logger,
+	}
+}
+
+func (c *RPCClient) Call(dc string, server *metadata.Server, method string, args, reply interface{}) error {
+	if !server.GRPCEnabled || !grpcAbleEndpoints[method] {
+		c.logger.Printf("[TRACE] Using RPC for method %s", method)
+		return c.rpcPool.RPC(dc, server.Addr, server.Version, method, server.UseTLS, args, reply)
+	}
+
+	conn, err := c.grpcConn(server)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Printf("[TRACE] Using GRPC for method %s", method)
+	return conn.Invoke(context.Background(), c.grpcPath(method), args, reply)
+}
+
+func (c *RPCClient) Ping(dc string, addr net.Addr, version int, useTLS bool) (bool, error) {
+	return c.rpcPool.Ping(dc, addr, version, useTLS)
+}
+
+func (c *RPCClient) Shutdown() error {
+	// Close the connection pool
+	c.rpcPool.Shutdown()
+	return nil
+}
+
+func (c *RPCClient) grpcConn(server *metadata.Server) (*grpc.ClientConn, error) {
+	host, _, _ := net.SplitHostPort(server.Addr.String())
+	addr := fmt.Sprintf("%s:%d", host, server.Port)
+
+	conn, ok := c.grpcConns.Load(addr)
+	if ok {
+		return conn.(*grpc.ClientConn), nil
+	}
+
+	co, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithDialer(dialGRPC))
+	if err != nil {
+		return nil, err
+	}
+
+	c.grpcConns.Store(addr, co)
+	return co, nil
```

Comment:
I wonder if it's possible to not re-invent grpc client's connection pool here?

Reply:
It's a bit obscure but the whole client-side loadbalancing thing is what I mean: https://godoc.org/google.golang.org/grpc/balancer

you still call dial but the "conn" returned is an abstract thing that will actually get requests loadbalanced to multiple servers according to your balancer and resolver methods you pass. This way we could recreate the retry during leader election and periodic load spreading we do in the current RPC layer which would be important for long-term adoption of gRPC.

Not asking you to do this unless you want to @Aestek , just leaving a note for myself since I looked :)

Reply:
yes I've looked into this as well but it was easier and safer to just reuse the mechanism already in place for classic RPC. But definitely something to look into in the future.

Reply:
I looked into this more and I agree this is a pretty complicated issue. For a start the grpc balancer interface and mechanism is pretty hard to understand and not relly documented well. The best place I found to understand it eventually was the original proposal doc for the most recent API change: https://github.com/menghanl/proposal/blob/8043d465114dfd9cba04ba96a87af6534e8a1519/L9-go-resolver-balancer-API.md

It would be possible I think to define custom resolver and balancer that replicate the current behaviour around server selection and get updated by serf etc. But it would be a pretty huge change in our router package.

Another subtlety is that the router.Manager currently uses `connPool.Ping` to check if a server is alive which is fine for now as we still have old RPCs in use and we can assume if server is up for one type of RPC it is for both, but eventually when we move to only gRPC we'd need to rethink that.

Even if we do all of that and make it work in an equivalent way, there is a bunch more thinking to put into Enterprise network area routing which comes into play for cross-DC calls between servers in different DCs.

gRPC also assumes that each RPC should be balanced between all available servers. While a custom balancer could change that, there are subtleties to consider for example typical balancer implementations initiate connections to _all_ servers and then just return the one that should be used. It's not clear to me if that means they open TCP connections to all of them eagerly or lazily but eagerly could change things a bunch for our users - maybe cause 5x more tcp connections to be opened to their servers.

gRPC also sends pings and actively managed connection status - failed connections are reported to the balancer and can be removed. It's not exactly a problem but working out how that meshes with our own Serf health for servers as handled by the Manager is not super clear to me.

As a final note, gRPC actively maintains all of the connections in the background so even if a balancer picked the same one consistently for each rebalance interval all of them would stay open. I _thought_ that was a behaviour change and that our connection pool closes unused one after rebalance but that doesn't seem to be the case from reading our code closely again so that should be OK.

tl;dr is that your approach here is a simpler way to keep all the existing logic around server selection unchanged for gRPC!